File: python/data_structures/segment_tree.py
class SegTree:
    def __init__(self, a_or_n, op=max, ident=-(10**30)):
        self.op = op
        self.ident = ident
        n0 = a_or_n if isinstance(a_or_n, int) else len(a_or_n)
        self.n = 1
        while self.n < n0:
            self.n <<= 1
        self.st = [ident] * (2 * self.n)
        if not isinstance(a_or_n, int):
            self.st[self.n:self.n + n0] = a_or_n
            for i in range(self.n - 1, 0, -1):
                self.st[i] = op(self.st[i << 1], self.st[i << 1 | 1])

    def set(self, i, value):
        i += self.n
        self.st[i] = value
        i >>= 1
        while i:
            self.st[i] = self.op(self.st[i << 1], self.st[i << 1 | 1])
            i >>= 1

    def query(self, l, r):
        left = self.ident
        right = self.ident
        l += self.n
        r += self.n
        while l < r:
            if l & 1:
                left = self.op(left, self.st[l])
                l += 1
            if r & 1:
                r -= 1
                right = self.op(self.st[r], right)
            l >>= 1
            r >>= 1
        return self.op(left, right)

File: cpp/data_structures/segment_tree.cpp
#include <bits/stdc++.h>
using namespace std;
template <class T, class Merge>
struct SegTree
{
  int n;
  T id;
  Merge merge;
  vector<T> st;
  SegTree(int n = 0, T id = T{}, Merge m = Merge()) : id(id), merge(m)
  {
    if (n)
      init(n);
  }
  void init(int N)
  {
    n = 1;
    while (n < N)
      n <<= 1;
    st.assign(2 * n, id);
  }
  void build(const vector<T> &a)
  {
    init(a.size());
    for (int i = 0; i < (int)a.size(); i++)
      st[n + i] = a[i];
    for (int i = n - 1; i; i--)
      st[i] = merge(st[i << 1], st[i << 1 | 1]);
  }
  void setPoint(int p, T v)
  {
    for (st[p += n] = v, p >>= 1; p; p >>= 1)
      st[p] = merge(st[p << 1], st[p << 1 | 1]);
  }
  T query(int l, int r)
  {
    T L = id, R = id;
    for (l += n, r += n; l <= r; l >>= 1, r >>= 1)
    {
      if (l & 1)
        L = merge(L, st[l++]);
      if (!(r & 1))
        R = merge(st[r--], R);
    }
    return merge(L, R);
  }
};
